nexus/src/app/instance_platform.rs

```diff
@@ -0,0 +1,481 @@
+/// serial number to produce a logical "disk ID" that those guests (and their
+/// users) may assume will remain stable for the lifetime of the disk.
+pub fn zero_padded_nvme_serial_from_str(s: &str) -> [u8; 20] {
+    let mut sn = [0u8; 20];
+
+    let bytes_from_slice = sn.len().min(s.len());
+    sn[..bytes_from_slice].copy_from_slice(&s.as_bytes()[..bytes_from_slice]);
+    sn
+}
+
+/// Describes a Crucible-backed disk that should be added to an instance
+/// specification.
+#[derive(Debug)]
+struct CrucibleDisk {
+    device_name: String,
+    device: ComponentV0,
+    backend: CrucibleStorageBackend,
+}
+
+/// Stores a mapping from Nexus disk IDs to Crucible disk descriptors. This
+/// allows the platform construction process to quickly determine the *device
+/// name* for a disk with a given ID so that that name can be inserted into the
+/// instance spec's boot settings.
+struct DisksById(BTreeMap<Uuid, CrucibleDisk>);
+
+impl DisksById {
+    /// Creates a disk list from an iterator over a set of disk and volume
+    /// records.
+    ///
+    /// The caller must ensure that the supplied `Volume`s have been checked out
+    /// (i.e., that their Crucible generation numbers are up-to-date) before
+    /// calling this function.
+    fn from_disks<'i>(
+        disks: impl Iterator<Item = (&'i db::model::Disk, &'i db::model::Volume)>,
+    ) -> Result<Self, Error> {
+        let mut map = BTreeMap::new();
+        for (disk, volume) in disks {
+            let slot = match disk.slot {
+                Some(s) => s.0,
+                None => {
+                    return Err(Error::internal_error(&format!(
+                        "disk {} is attached but has no PCI slot assignment",
+                        disk.id()
+                    )));
+                }
+            };
+
+            let pci_path = get_pci_bdf(slot, PciDeviceKind::Disk)?;
+            let device = ComponentV0::NvmeDisk(NvmeDisk {
+                backend_id: SpecKey::Uuid(disk.id()),
+                pci_path,
+                serial_number: zero_padded_nvme_serial_from_str(
+                    &disk.name().to_string(),
+                ),
+            });
+
+            let backend = CrucibleStorageBackend {
+                readonly: false,
+                request_json: volume.data().to_owned(),
+            };
+
+            let device_name = component_names::device_name_from_id(&disk.id());
+            if map
+                .insert(
+                    disk.id(),
+                    CrucibleDisk { device_name, device, backend },
+                )
+                .is_some()
+            {
+                return Err(Error::internal_error(&format!(
+                    "instance has multiple attached disks with ID {}",
+                    disk.id()
+                )));
+            }
+        }
+
+        Ok(Self(map))
+    }
+}
+
+/// A list of named components to add to an instance's spec.
+struct Components(HashMap<String, ComponentV0>);
```

Comment:
i notice that we use `HashMap` here but `BTreeMap` for `DisksById`, any rationale for that?

Reply:
`sled_agent_client::types::InstanceSpecV0`'s `components` member is a `HashMap` because that's what Progenitor uses for map types. I usually reach for `BTreeMap` by default and then switch it when something breaks. I'll leave a note here.